patch/llpcPatchResourceCollect.cpp

```diff
@@ -2372,6 +2402,385 @@ void PatchResourceCollect::MapGsBuiltInOutput(
+void PatchResourceCollect::ReassembleOutputExportCalls()
+{
+    if (m_inOutCalls.empty())
+    {
+        return;
+    }
+
+    auto& inOutUsage = m_pContext->GetShaderResourceUsage(m_shaderStage)->inOutUsage;
+
+    // Collect the components of a vector exported from each packed location
+    // Assume each location exports a vector with four components
+    std::vector<std::array<Value*, 4>> packedComponents(m_inOutCalls.size());
+    for (auto pCall : m_inOutCalls)
+    {
+        InOutLocation origOutLoc = {};
+        origOutLoc.locationInfo.location = cast<ConstantInt>(pCall->getOperand(0))->getZExtValue();
+        origOutLoc.locationInfo.component = cast<ConstantInt>(pCall->getOperand(1))->getZExtValue();
+        origOutLoc.locationInfo.half = false;
+
+        const InOutLocation* pNewInLoc = nullptr;
+        const bool isFound = m_pLocationMapManager->FindMap(origOutLoc, pNewInLoc);
+        if (isFound == false)
+        {
+            continue;
+        }
+
+        auto& components = packedComponents[pNewInLoc->locationInfo.location];
+        components[pNewInLoc->locationInfo.component] = pCall->getOperand(2);
+    }
+
+    // Re-assamble XX' output export calls for each packed location
+    IRBuilder<> builder(*m_pContext);
+    builder.SetInsertPoint(m_inOutCalls.back());
+
+    auto& outputLocMap = inOutUsage.outputLocMap;
+    outputLocMap.clear();
+
+    Value* args[3] = {};
+    uint32_t consectiveLocation = 0;
+    for (auto components : packedComponents)
+    {
+        uint32_t compCount = 0;
+        for (auto pComp : components)
+        {
+            if (pComp != nullptr)
+            {
+                ++compCount;
+            }
+        }
+
+        if (compCount == 0)
+        {
+            break;
+        }
+
+        // Construct the output vector
+        Value* pOutValue = (compCount == 1) ? components[0] :
+                           UndefValue::get(VectorType::get(m_pContext->FloatTy(), compCount));
+        for (auto compIdx = 0; compIdx < compCount; ++compIdx)
+        {
+            // Type conversion from non-float to float
+            Value* pComp = components[compIdx];
+            Type* pCompTy = pComp->getType();
+            if (pCompTy->isIntegerTy())
+            {
+                // i8/i16 -> i32
+                if (pCompTy->getScalarSizeInBits() < 32)
+                {
+                    pComp = builder.CreateZExt(pComp, m_pContext->Int32Ty());
+                }
+                // i32 -> float
+                pComp = builder.CreateBitCast(pComp, m_pContext->FloatTy());
+            }
+            else if (pCompTy->isHalfTy())
+            {
+                // f16 -> float
+                pComp = builder.CreateFPExt(pComp, m_pContext->FloatTy());
+            }
+
+            if (compCount > 1)
+            {
+                Value* pCompIdx = ConstantInt::get(m_pContext->Int32Ty(), compIdx);
+                pOutValue = builder.CreateInsertElement(pOutValue, pComp, pCompIdx);
```

Comment:
You can just pass compIdx directly to CreateInsertElement (there is an overload that takes an integer).

Reply:
Thanks. Fixed.